File: 5/C++/main.cpp
/**
* @author Matias Grioni
* @date 01:10 20/3/15
* @desc Fifth Project Euler problem. Finds smallest number
*       that is divisible by all numbers less than some
*       inputed number.
*/

#include <iostream>

using namespace std;

int gcd(int a, int b) {
    int maxNum = (a > b) ? a : b;
    int result = 1;

    for(int i = 1; i <= maxNum; i++) {
        if(a % i == 0 && b % i == 0)
            result = i;
    }

    return result;
}

int lcm(int a, int b) {
    return (a * b) / gcd(a, b);
}

int lcmRange(int limit) {
    if(limit == 1) {
        return 1;
    } else {
        return lcm(limit, lcmRange(limit - 1));
    }
}

int main()
{
    int limit;
    cout << "Find number divisible by all numbers less than > ";
    cin >> limit;

    cout << "Answer is: " << lcmRange(limit);

    return 0;
}

File: 5/Haskell/main.hs
{-
 - Author: Matias Grioni
 - Created: 7/7/2015
 -
 - Finds the least common multiple of all numbers between 1 and n.
 -}

module Main where

gcd' :: Int -> Int -> Int
gcd' a 0 = a
gcd' a b = gcd b $ a `div` b

lcm' :: Int -> Int -> Int
lcm' a b = a * b `div` gcd a b

main = do
    putStrLn "Find the least common multiple for all numbers between 1 and n"
    putStrLn "Enter n > "
    input <- getLine
    let n = read input :: Int

    print $ foldr lcm' 1 [1..n]
